algorithmic_efficiency/spec.py

```diff
@@ -422,6 +423,7 @@ def update_params(workload: Workload,
                   batch: Dict[str, Tensor],
                   loss_type: LossType,
                   optimizer_state: OptimizerState,
+                  train_state: Dict[str, Any],
```

Comment:
Can we make this an optional kwarg and default it to None for backwards compatibility?

Reply:
train_state is just a small dict[str, float] right? does it take a significant amount of time to copy it?

Also, good point about checking before passing the arg to the train_state. I guess the code may look a little weird to people who don't have context about the backward compatiliblity. But I can't think of a cleaner solution that also maintains the backwards compatibility, so let's add your suggestion.
Alternatively we could make a small change to all submissions and add **kwargs to update_params. But I feel like that is a slippery slope. WDYT?

Reply:
Yes, `train_state` is very small, it should add a minimal overhead indeed, I just didn't like the idea of having extra ops in the submission time 😅. But we could also pass a `MappingProxyType(train_state)` to pass a read-only version without need for copying, which should be faster.

Regarding the last point, I agree with you. We can start like this and perhaps if we later want to simplify the logic at the cost of adding **kwargs, we can do that.

Reply:
Ok, let's go ahead with the `inspect` route and I will merge this PR in

Reply:
Ok, I changed it, and it should be good to go now!